### PR TITLE
add `--filter` and package pattern arguments to `bun outdated`

### PR DESCRIPTION
**Pull request by dylan-conway**
### What does this PR do?
- Adds workspace filtering to `bun outdated` through `--filter`. You can glob for workspace package names or workspace paths. Examples:
```bash
# Path patterns:
# Matches all workspaces in `./packages`
bun outdated --filter="./packages/*"

# Name patterns:
# All workspaces with names starting with `pkg`
bun outdated --filter="pkg*"

# All workspaces excluding `pkg1`
bun outdated --filter="*" --filter="!pkg1"
```
- Adds dependency name filtering. Works the same as workspace name patterns, and can be used in combination with `--filter`. Example:
```bash
# All dependencies starting with `is-` (could display `is-even`, `is-odd`, ...)
bun outdated "is-*"

# With `--filter`. Matches all outdated `jquery` dependencies in all workspaces.
bun outdated jquery --filter="*"
```
- fixes a bug that would occur when printing versions with tags in the `bun outdated` table

### How did you verify your code works?
Added tests for `--filter` and args

### CODE REVIEW

**Comment by robobun** [comment]
@dylan-conway, your commit fcbc645ffe9da0bcbc572a0718fb051cd0aaf9fc is building: [`#2386`](https://buildkite.com/bun/bun/builds/2386)

File: src/cli/outdated_command.zig [hunk 1]
```diff
@@ -180,92 +186,282 @@ pub const OutdatedCommand = struct {
...
+
+            const patterns_buf = bun.default_allocator.alloc([]const u32, args.len) catch bun.outOfMemory();
+            for (args, patterns_buf) |arg, *converted| {
+                if (arg.len == 0) {
+                    converted.* = &.{};
+                    continue;
+                }
+
+                const length = bun.simdutf.length.utf32.from.utf8.le(arg);
+                const convert_buf = bun.default_allocator.alloc(u32, length) catch bun.outOfMemory();
```
Comment (Jarred-Sumner) [inline src/cli/outdated_command.zig:338]: seems crazy to me that we have to drop down to simdutf here

File: src/cli/outdated_command.zig [hunk 2]
```diff
@@ -180,92 +186,282 @@ pub const OutdatedCommand = struct {
...
+                        if (filter.len == 0) continue;
+                        const res = pkg_resolutions[workspace_pkg_id];
+
+                        const res_path = switch (res.tag) {
+                            .workspace => res.value.workspace.slice(string_buf),
+                            .root => FileSystem.instance.top_level_dir,
+                            else => unreachable,
+                        };
+
+                        const abs_res_path = path.joinAbsString(FileSystem.instance.top_level_dir, &[_]string{res_path}, .posix);
```
Comment (Jarred-Sumner) [inline src/cli/outdated_command.zig:289]: probably should use joinAbsStringBuf incase something else is making use of the shared string buffer?

File: src/cli/outdated_command.zig [hunk 3]
```diff
@@ -180,92 +186,282 @@ pub const OutdatedCommand = struct {
...
+        defer {
+            for (converted_filters) |converted| {
+                const filter, _ = converted;
+                allocator.free(filter);
+            }
+            allocator.free(converted_filters);
+        }
+
+        // move all matched workspaces to front of array
+        var i: usize = 0;
```
Comment (Jarred-Sumner) [inline src/cli/outdated_command.zig:271]:
```suggestion
        var i: usize = 0;
        var abs_res_path_buf: bun.PathBuffer = undefined;
```

File: src/cli/outdated_command.zig (hunk 2 above)
Comment (Jarred-Sumner) [inline src/cli/outdated_command.zig:289]:
```suggestion
                        const abs_res_path = path.joinAbsStringBuf(&abs_res_path_buf, FileSystem.instance.top_level_dir, &[_]string{res_path}, .posix);
```

**Comment by Jarred-Sumner** [comment]
This is mergeable so I'm going to merge it

I do think that we should have some canonical way to "get a validated UTF-8 string from bytes" that doesn't directly involve simdutf

**Pull request merged**